### PR TITLE
openssl_certificate: Sign with your own CA

### PR DESCRIPTION
**Pull request by psykotox**
##### SUMMARY

Add new provider `ownca` that can be used to sign with your own CA certificate. You Ca certificate can be a self-signed certificate. Very useful in private network when you want to use https on local domain with a private CA certificate installed on all devices.
I added also 2 new features:
- renew certificate if csr change
- renew certificate before its expiration (`remaining_days`)

##### ISSUE TYPE

 - Feature Pull Request
 
##### COMPONENT NAME

openssl_certificate

##### ANSIBLE VERSION

```
ansible 2.5.0
  config file = None
  configured module search path = [u'/home/loic/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.13 (default, Nov 24 2017, 17:33:09) [GCC 6.3.0 20170516]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by MarkusTeufelberger** [comment]
For the future: Please don't add multiple unrelated features in one PR (in this case: `remaining_days`, csr validation and the `ownca` provider)

Do you really store your CA private keys on client machines? This maybe should be an action plugin...

**Comment by psykotox** [comment]
I don't store CA private keys on client machines but on a server that used like a CA server.
When I want to sign a certificate for a new local domain I use openssl_certificate on that server and I transfer the signed certificate on the good machine.

What do you mean with "action plugin" ?

**Comment by MarkusTeufelberger** [comment]
http://docs.ansible.com/ansible/devel/plugins/action.html

**Comment by psykotox** [comment]
If I understand correctly, the `action plugin` are executed on the master and the modules on the remote machine. So you suggest to keep the private CA key on the master and sign the certificate on the master and then transfer it with an `action plugin` ?
Personally I prefer to run `openssl_certificate`with `delegate_to` to choose my key server and then make a copy on my remote machine.

**Comment by MarkusTeufelberger** [comment]
Yeah, that's also an option (especially if you want to decouple CA machine, master and target host), it probably should be spelled out explicitly though that `ownca` should NOT run on a target machine, but on a dedicated CA machine, then using `fetch` or whatever the signed certificate gets moved to its final destination.

**Comment by psykotox** [comment]
I just added a note about that ;)

**Comment by michaelkaye** [comment]
Hi there,

I was just 2/3 of the way through writing a feature request to ask for this before finding this PR.

I have the CA use case myself - I want to use a CA stored on my infrastructure machine that runs ansible to sign CSRs generated on the target machines - using fetch/copy to move the CSR & certificate back without exposing the private key outside of the target machine. I like this and think it's what I need to simplify my tasks from having to do a command invocation of `openssl x509`

However, I'm not as sure about the `remaining_days` or csr validation features in this PR; the logic needed for deciding to renew a CSR isn't just within the task running openssl_certificate - the check actually wants to happen on the target machine, but the openssl_certificate will be happening on an infrastructure machine, so is probably fine to no longer have the CSRs available on it . The sort of logic you want is:

```
- test target machine to see if certificate expires soon or is missing, or private key is missing
when test fails do all the following:
  - generate private key
  - create CSR
  - fetch CSR to CA machine
  - sign CSR with CA
  - copy certificate back to target
```
So i think this test is better off being a block of tasks filtered by a when: rather than being tied to the signing that may be on an aribtary other machine.

Personally, I'd love to see this split in two and the CA signing bit merged; but further thought on the new tests that we're adding to this module.

**Comment by k0ste** [comment]
I found this PR because on way moving from shell to module-based cert issuing.
I use own CA for a years with ansible, of course your CA should be in some dedicated place. I make task that generate key,csr,crt with `delegate_to`. After tihs next task is fetching this files to execution node. It's easy.

**Comment by zlbabe** [comment]
What are we waiting for to merge this PR ?

**Comment by MarkusTeufelberger** [comment]
I disagree with the (unrelated) `remaining_days` parameter for example and I agree with michael's assessment and conclusion.

**Comment by Spredzy** [comment]
@psykotox thanks for this PR and sorry for missing the 2.5 window on this one.

Could you : 

  * Rebase all the commits in a single commit
  * Add a test scenario for this PR

Also I agree with @michaelkaye comment. I think that the condition for re-signing isn't only based on `remaining_days` and hence the logic should be kept outside of the scope of the `openssl_certificate` module or re think in some way.

It is fairly easy with ansible to retrieve a list of ssl certificatse to re-sign (based on a specific condition) and fetch them to `openssl_certificate` for signing.

As some users have shown interest in this feature,  I am looking forward to integrate it.

**Comment by psykotox** [comment]
If I understand correctly @michaelkaye @MarkusTeufelberger @Spredzy, you suggest: do not use `remaining_days` on the CA machine and do not keep the generated certificate on the CA machine. But check the generated certificate on the target machine with `assertonly` and run the re-singing (`delegate_to` CA machine) filtered with `when` conditions ?

I agree with that suggestion.

If it's Ok for you, I remove the parameter `remaining_days` and I begin to add test scenario.

**Comment by michaelkaye** [comment]
Hi @psykotox - Yes, that's what I was looking for with this module.

The key thing is that we don't make the module force one way to do things - we don't want it to force the CA to be on a different machine nor do we want to force it to be on the same machine. Making a nice toolbox we can use in either way is what I'd love to see.

**Comment by zlbabe** [comment]
Is there any way to speed up the merge of this PR cause I really need it ? let me know if I can help

**Comment by MarkusTeufelberger** [comment]
It'll make it into Ansible 2.6 at the earliest (since it changes API), so if you need it before that, you'll have to temporarily maintain a local copy anyways. You could just take the code from this PR for example, or you could help by splitting the various features that are introduced in here and submit them separately.

**Comment by psykotox** [comment]
I don't have much time for my PR right now. 
But I'll try to do that ASAP:
- Remove `remaining_days`
- Add a test scenario for this PR
- Rebase all the commits in a single commit

**Comment by psykotox** [comment]
done
@MarkusTeufelberger @Spredzy the PR is ready to ship

File: lib/ansible/modules/crypto/openssl_certificate.py [hunk 1]
```diff
@@ -847,7 +1031,7 @@ def main():
 
     if not pyopenssl_found:
         module.fail_json(msg='The python pyOpenSSL library is required')
-    if module.params['provider'] in ['selfsigned', 'assertonly']:
+    if module.params['provider'] in ['selfsigned', 'ownca' 'assertonly']:
```
Comment (MarkusTeufelberger) [inline lib/ansible/modules/crypto/openssl_certificate.py]: misses a comma after `'ownca'`

Comment (psykotox) [reply lib/ansible/modules/crypto/openssl_certificate.py]: Yes, I fix it

**Comment by MarkusTeufelberger** [comment]
Looks great from skimming over it, thanks a lot for your contribution.

As far as I understand it, a `ownca` openssl_certificate will in most parts sign whatever the csr wants. This is also the case with `selfsigned`, but not ideal in both cases (in many cases there probably should be a very thorough `assertonly` run afterwards). That's not 100% related to this PR though, so from my side it is good to go.

**Comment by MarkusTeufelberger** [comment]
shipit

**Comment by thomwiggers** [comment]
shipit

**Comment by ghost** [comment]
This feature is a lifesaver for large internal deployments. It looks like the option definitions just need a version bump. Happy to open a new PR if OP has moved on

**Comment by MarkusTeufelberger** [comment]
shipit

**Comment by Spredzy** [comment]
shipit

(Even though we will need to re: shipit after version bump)

**Comment by MarkusTeufelberger** [comment]
It already is written for version 2.7

I guess it is not merged automatically because this PR is touching several files. I had a similar issue and asked on IRC to have someone take a look and merge manually.

**Comment by Spredzy** [comment]
@MarkusTeufelberger just asked on IRC. Hopefully will be merged today.

**Comment by redbaron** [comment]
In case anybody else stumbled upon this and scratches their head how to set not_after to something else, but 10 years, here is what I came up with:

1. `pip install jinja2-time`
2. add `jinja2_extensions = jinja2_time.TimeExtension` to your `ansible.cfg`'s `[default]` section
3. use `ownca_not_after: "{% now 'utc' + 'days=365', '%Y%m%d%H%M%SZ' %}"`

**Pull request merged**